pkg/agent/cniserver/interface_configuration_linux.go

```diff
@@ -0,0 +1,342 @@
+}
+
+func parseContainerIfaceFromResults(cfgArgs *cnipb.CniCmdArgs, prevResult *current.Result) *current.Interface {
+	for _, intf := range prevResult.Interfaces {
+		if intf.Name == cfgArgs.Ifname {
+			return intf
+		}
+	}
+	return nil
+}
+
+func (ic *ifConfigurator) checkContainerInterface(
+	containerNetns, containerID string,
+	containerIface *current.Interface,
+	containerIPs []*current.IPConfig,
+	containerRoutes []*cnitypes.Route) (*vethPair, error) {
+	var contVeth *vethPair
+	// Check netns configuration
+	if containerNetns != containerIface.Sandbox {
+		klog.Errorf("Sandbox in prevResult %s doesn't match configured netns: %s",
+			containerIface.Sandbox, containerNetns)
+		return nil, fmt.Errorf("sandbox in prevResult %s doesn't match configured netns: %s",
+			containerIface.Sandbox, containerNetns)
+	}
+	netns, err := ns.GetNS(containerNetns)
+	if err != nil {
+		klog.Errorf("Failed to check netns config %s: %v", containerNetns, err)
+		return nil, err
+	}
+	defer netns.Close()
+	// Check container interface configuration
+	if err := netns.Do(func(netNS ns.NetNS) error {
+		var errlink error
+		// Check container link config
+		contVeth, errlink = validateContainerInterface(containerIface)
+		if errlink != nil {
+			return errlink
+		}
+		// Check container IP config
+		if err := ip.ValidateExpectedInterfaceIPs(containerIface.Name, containerIPs); err != nil {
+			return err
+		}
+		// Check container route config
+		if err := ip.ValidateExpectedRoute(containerRoutes); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		klog.Errorf("Failed to check container %s interface configurations in netns %s: %v",
+			containerID, containerNetns, err)
+		return nil, err
+	}
+	return contVeth, nil
+}
+
+func validateContainerInterface(intf *current.Interface) (*vethPair, error) {
+	link, err := validateInterface(intf, true)
+	if err != nil {
+		return nil, err
+	}
+
+	linkName := link.Attrs().Name
+	veth := &vethPair{}
+	_, veth.peerIndex, err = ip.GetVethPeerIfindex(linkName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get veth peer index for veth %s: %v", linkName, err)
+	}
+	veth.ifIndex = link.Attrs().Index
+	if intf.Mac != link.Attrs().HardwareAddr.String() {
+		return nil, fmt.Errorf("interface %s MAC %s doesn't match container MAC: %s",
+			intf.Name, intf.Mac, link.Attrs().HardwareAddr.String())
+	}
+	veth.name = linkName
+	return veth, nil
+}
+
+func (ic *ifConfigurator) validateContainerPeerInterface(interfaces []*current.Interface, containerVeth *vethPair) (*vethPair, error) {
+	// Iterate all the passed interfaces and look up the host interface by
+	// matching the veth peer interface index.
+	for _, hostIntf := range interfaces {
+		if hostIntf.Sandbox != "" {
+			// Not in the default Namespace. Must be the container interface.
+			continue
+		}
+		link, err := validateInterface(hostIntf, false)
+		if err != nil {
+			klog.Errorf("Failed to validate interface %s: %v",
+				hostIntf.Name, err)
```

Comment:
nit: feel this line doesn't really need to be wrapped as there are lines longer than it in the same screen.